src/main/java/org/javarosa/core/util/DAG.java

```diff
@@ -138,6 +142,33 @@ public Stack<I> getSinks() {
         return roots;
     }
 
+    public Set<I> getRelatedRecords(Set<I> recordIds) {
+        Set<I> visited = new HashSet<>();
+        LinkedList<I> queue = new LinkedList<>(recordIds);
+        while (!queue.isEmpty()) {
+            I current = queue.poll();
+            if(visited.contains(current)){
+                continue;
+            }
+            visited.add(current);
+            addNeighbors(edges, current, queue, visited);
+            addNeighbors(inverseEdges, current, queue, visited);
+        }
+        return visited;
+    }
+
+    private void addNeighbors(Hashtable<I, Vector<Edge<I, E>>> edges, I current, LinkedList<I> queue,
```

Comment:
"Add" is likely to be a misleading name for this method. As stated it sounds like it would be adding neighbors to the DAG itself. 

I'd probably just rename this to `enqueueNeighbors` or `accumulateNeighbors`, but it should be clear that the method itself is stateless.

Reply:
https://github.com/dimagi/commcare-core/pull/1460/commits/24da04f8e80bb90348f69698c9826ba0a2063f48